pipeline/transform.py

```diff
@@ -0,0 +1,211 @@
+
+    return sales_info
+
+
+def get_genres_from_url(artist_url: str) -> list[str]:
+    """
+    Gets the lists of genres from a given artist URL.
+    Returns an empty list if none can be found or an error occurs.
+    """
+    try:
+        full_url = convert_to_full_url(artist_url)
+        response = requests.get(full_url, timeout=1000)
+
+        if response.status_code == 200:
+            logging.info("Getting genre tags for %s", full_url)
+            soup = BeautifulSoup(response.text, 'html.parser')
+            tags = soup.find_all('a', class_='tag')
+
+            if tags:
+                genres = [tag.text for tag in tags]
+                return genres
+
+            logging.info("No genres listed for %s", full_url)
+            return []
+
+        logging.warning("Failed to connect to %s, Status Code: %s",
+                        full_url, response.status_code)
+        return []
+
+    except requests.exceptions.RequestException as e:
+        logging.error("Request failed for %s: %s", artist_url, e)
+
+    except Exception as e:
+        logging.error("An unexpected error occurred: %s", e)
+
+    return []
+
+
+def get_release_date_from_url(artist_url: str) -> str:
+    """
+    Retrieves the release date of a song or album provided in the URL.
+    Returns an empty string if it cannot be found or an error occurs.
+    """
+    try:
+        full_url = convert_to_full_url(artist_url)
+        response = requests.get(full_url, timeout=1000)
+
+        if response.status_code == 200:
+            logging.info("Getting release date for %s", full_url)
+            soup = BeautifulSoup(response.text, 'html.parser')
+            meta_tag = soup.find('meta', attrs={'name': 'description'})
+
+            if meta_tag:
+                description_content = meta_tag.get('content')
+                if description_content:
+                    date_pattern = r'released\s(\d{1,2}\s[a-zA-Z]+\s\d{4})'
+                    match = re.search(date_pattern, description_content)
+                    if match:
+                        release_date = match.group(1)
+                        return convert_date_format(release_date)
+                else:
+                    logging.info(
+                        "No release date found in the meta description for %s", full_url)
+        else:
+            logging.warning("Failed to connect to %s, Status Code: %s",
+                            full_url, response.status_code)
+
+    except requests.exceptions.RequestException as e:
+        logging.error("Request failed for %s: %s", artist_url, e)
+
+    except Exception as e:
+        logging.error("An unexpected error occurred: %s", e)
+
+    return ""
+
+
+def create_sales_dataframe(sales_info: list[dict]) -> pd.DataFrame:
+    """
+    Inserts given sales data into a dataframe and returns it.
+    Also exports it as a .csv file.
+    """
+    sales_df = pd.json_normalize(sales_info)
+
+    sales_df.to_csv("MUSIC_DATA.csv")
+
```

Comment:
Could we add the sale year and month to the dataframe just for the trend analysis for the dashboard?

df["sale_year"] = df["sale_date"].dt.year
df["sale_month"] = df["sale_date"].dt.month